plugin/pulpcore/plugin/download/asyncio/base.py

```diff
@@ -0,0 +1,153 @@
+from collections import namedtuple
+import hashlib
+import os
+import tempfile
+
+from pulpcore.app.models import Artifact
+from .exceptions import DigestValidationError, SizeValidationError
+
+
+DownloadResult = namedtuple('DownloadResult', ['url', 'artifact_attributes', 'path'])
+"""
+Args:
+    url (str): The url corresponding with the download.
+    path (str): The absolute path to the saved file
+    artifact_attributes (dict): Contains keys corresponding with pulpcore.plugin.Artifact fields.
+        This includes the computed digest values along with size information.
+"""
+
+class BaseDownloader(object):
+    """
+    The base class of all downloaders. This is an abstract class and is meant to be subclassed.
+
+    This provides data digest calculation and validation and the writing to a file.
+
+    All subclassed downloaders should pass all downloaded data the into
+    :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data` to allow the file digest to
+    be computed while data is written to disk. This avoids having to re-read the data later. The
+    digests computed are required to save the file as an Artifact, so we need to compute them.
+
+    The :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data` method by default
+    writes to a random file in the current working directory or you can pass in your own file
+    object. See the ``custom_file_object`` keyword argument for more details. Allowing the download
+    instantiator to define the file to receive data allows the streamer to receive the data instead
+    of having it written to disk.
+    """
+
+    def __init__(self, custom_file_object=None, expected_digests=None, expected_size=None):
+        """
+        Create a BaseDownloader object. This is expected to be called by all subclasses.
+
+        Args:
+            custom_file_object (file object): An open, writable file object that downloaded data
+                can be written to by
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+            expected_digests (dict): Keyed on the algorithm name provided by hashlib and stores the
+                value of the expected digest. e.g. {'md5': '912ec803b2ce49e4a541068d495ab570'}
+            expected_size (int): The number of bytes the download is expected to have.
+        """
+        if custom_file_object:
+            self._writer = custom_file_object
```

Comment:
Looks like `self.path` only exists when `custom_file_object` is not passed to `__init__()`.  Suggest adding: `self.path = None` to ensure the instance always has the attribute.

Reply:
Yeah good catch. Fixed in next push along w/ some doc lines explaining the cases of None versus when its actually a path.